### ISSUE TITLE
Adjust readiness checks for new "Improved Health endpoint in Admin API"

### ISSUE THREAD

**Issue by joejulian** [redpanda-data/helm-charts#515]
### What would you like to be added?

https://github.com/redpanda-data/planning/issues/39 is going to change the health endpoint. Change the readiness checks to ensure rolling pods is properly blocked to prevent outages and data loss.

### Why is this needed?

JIRA Link: [K8S-40](https://redpandadata.atlassian.net/browse/K8S-40)

[K8S-40]: https://redpandadata.atlassian.net/browse/K8S-40?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by joejulian** [comment]
Blocked waiting on core changes

**Comment by aneagoe** [comment]
@joejulian The link seems to be broken (results in 404).
The current implementation of the readiness check is disruptive. As soon as one pod enters terminating state, the command returns cluster health false. This in turn disables any traffic to the pod. Because cluster health is actually expected to be degraded during pod rollouts, all members of the cluster then fail their readiness check. In turn, k8s services refuse to forward traffic to them. So, external services (whether NodePort or LoadBalancer) then fail altogether, even though the cluster is fine.
Until this is addressed, is there any other healthcheck that could be used? Additionally, perhaps an option to customize the healthcheck?

**Comment by joejulian** [comment]
That's a bug. Thanks for bringing it to our attention.

https://github.com/redpanda-data/helm-charts/blob/ddb1d1d52b4973796b4572296e57ac11a6c73edc/charts/redpanda/templates/service.internal.yaml#L32
Is added to prevent that from happening internally but never was added to the external services.

Created a new issue for that: https://github.com/redpanda-data/helm-charts/issues/543

**Comment by joejulian** [comment]
Because of the nature of the Kafka protocol, the state of the data has a greater impact on readiness than that of a broker. The protocol, itself, handles brokers being unavailable so we don't need kubernetes managing that. 

Where readiness really comes into play is satisfying the pod distribution budget during restarts. The topics need to have leaders and should not be under replicated. This is the signal we use for readiness to ensure the ability to safely allow pod deletion.

**Comment by aneagoe** [comment]
Hrm... but what about bootstrap service? Client needs to be able to hit a healthy broker that would in turn return the addresses of all the healthy brokers in the cluster. For individual broker service, the above holds true, it's not relevant. 
I've tested with `kcat` from outside the cluster during rolling updates, for the most part it handles it correctly (I do create my custom services with `publishNotReadyAddresses: true`). But there are some cases where kcat bails out as the service tries to redirect requests to the pod that is in `Terminating` state. It seems that once `publishNotReadyAddresses` is set to true, the pod state is also ignored.

**Comment by david-yu** [comment]
Given that this very old, we will close issue. If there are new issues that need to be addressed please file a separate issue as a follwup.
